File: UCSample/UCSample/SettingsViewController.m
//
//  SettingsViewController.m
//  UCSample
//
//  Copyright (c) 2013 PortSIP Solutions, Inc. All rights reserved.
//

#import "SettingsViewController.h"
#import "SettingCell.h"
#include "AppDelegate.h"

#define kNumbersOfSections 3

static NSString *kAudioCodecsKey = @"Audio Codecs";
static NSString *kVideoCodecsKey = @"Video Codecs";
static NSString *kAdvanceFeatureKey = @"Advance Feature";
@interface SettingsViewController ()

@end

@implementation SettingsViewController

- (id)initWithStyle:(UITableViewStyle)style {
  self = [super initWithStyle:style];
  if (self) {
    // Custom initialization
  }
  return self;
}

- (void)viewDidLoad {
  [super viewDidLoad];

  // Uncomment the following line to preserve selection between presentations.
  // self.clearsSelectionOnViewWillAppear = NO;

  // Uncomment the following line to display an Edit button in the navigation
  // bar for this view controller.
  // self.navigationItem.rightBarButtonItem = self.editButtonItem;

  settingsAudioCodec = [NSMutableArray arrayWithCapacity:20];
  SettingItem *item = [[SettingItem alloc] init];
  item.index = 0;
  item.name = @"OPUS";
  item.enable = YES;
  item.codeType = AUDIOCODEC_OPUS;
  [settingsAudioCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 1;
  item.name = @"G.729";
  item.enable = YES;
  item.codeType = AUDIOCODEC_G729;
  [settingsAudioCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 2;
  item.name = @"PCMA";
  item.enable = YES;
  item.codeType = AUDIOCODEC_PCMA;
  [settingsAudioCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 3;
  item.name = @"PCMU";
  item.enable = YES;
  item.codeType = AUDIOCODEC_PCMU;
  [settingsAudioCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 4;
  item.name = @"GSM";
  item.enable = NO;
  item.codeType = AUDIOCODEC_GSM;
  [settingsAudioCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 5;
  item.name = @"G.722";
  item.enable = NO;
  item.codeType = AUDIOCODEC_G722;
  [settingsAudioCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 6;
  item.name = @"iLBC";
  item.enable = NO;
  item.codeType = AUDIOCODEC_ILBC;
  [settingsAudioCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 7;
  item.name = @"AMR";
  item.enable = NO;
  item.codeType = AUDIOCODEC_AMR;
  [settingsAudioCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 8;
  item.name = @"AMRWB";
  item.enable = NO;
  item.codeType = AUDIOCODEC_AMRWB;
  [settingsAudioCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 9;
  item.name = @"SpeexNB(8Khz)";
  item.enable = NO;
  item.codeType = AUDIOCODEC_SPEEX;
  [settingsAudioCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 10;
  item.name = @"SpeexWB(16Khz)";
  item.enable = NO;
  item.codeType = AUDIOCODEC_SPEEXWB;
  [settingsAudioCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 11;
  item.name = @"ISACWB(16Khz)";
  item.enable = NO;
  item.codeType = AUDIOCODEC_ISACWB;
  [settingsAudioCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 12;
  item.name = @"ISACSWB(32Khz)";
  item.enable = NO;
  item.codeType = AUDIOCODEC_ISACSWB;
  [settingsAudioCodec addObject:item];

  // Video codec item
  settingsVideoCodec = [NSMutableArray arrayWithCapacity:10];

  item = [[SettingItem alloc] init];
  item.index = 101;
  item.name = @"H.264";
  item.enable = YES;
  item.codeType = VIDEO_CODEC_H264;
  [settingsVideoCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 102;
  item.name = @"VP8";
  item.enable = NO;
  item.codeType = VIDEO_CODEC_VP8;
  [settingsVideoCodec addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 103;
  item.name = @"VP9";
  item.enable = NO;
  item.codeType = VIDEO_CODEC_VP9;
  [settingsVideoCodec addObject:item];

  // Advance Feature
  settingsAdvanceFeature = [NSMutableArray arrayWithCapacity:10];
  if ([[[UIDevice currentDevice] systemVersion] floatValue] >= 10.0) {
    item = [[SettingItem alloc] init];
    item.index = 300;
    item.name = @"Integrated Calling";
    item.enable = shareAppDelegate.callManager.enableCallKit;
    item.codeType = -1;
    [settingsAdvanceFeature addObject:item];
  }

  item = [[SettingItem alloc] init];
  item.index = 301;
  item.name = @"Push Notification";
  item.enable = shareAppDelegate.enablePushNotification;
  item.codeType = -1;
  [settingsAdvanceFeature addObject:item];

  item = [[SettingItem alloc] init];
  item.index = 302;
  item.name = @"Force Background";
  item.enable = shareAppDelegate.enableForceBackground;
  item.codeType = -1;
  [settingsAdvanceFeature addObject:item];
}

- (void)didReceiveMemoryWarning {
  [super didReceiveMemoryWarning];
  // Dispose of any resources that can be recreated.
}

- (void)viewWillDisappear:(BOOL)animated {
  // save change
  [portSIPSDK clearAudioCodec];

  for (SettingItem *item in settingsAudioCodec) {
    if (item.enable) {
      [portSIPSDK addAudioCodec:(AUDIOCODEC_TYPE)item.codeType];
    }
  }

  [portSIPSDK clearVideoCodec];
  for (SettingItem *item in settingsVideoCodec) {
    if (item.enable) {
      [portSIPSDK addVideoCodec:(VIDEOCODEC_TYPE)item.codeType];
    }
  }

  // Save Value
  NSUserDefaults *settings = [NSUserDefaults standardUserDefaults];
  for (SettingItem *item in settingsAdvanceFeature) {
    switch (item.index) {
    case 300: // Integrated Calling
      shareAppDelegate.callManager.enableCallKit = item.enable;
      [settings setBool:item.enable forKey:@"CallKit"];
      break;
    case 301: // Push Notification
      if (shareAppDelegate.enablePushNotification != item.enable) {
        shareAppDelegate.enablePushNotification = item.enable;
        [shareAppDelegate refreshPushStatusToSipServer:item.enable];
      }

      [settings setBool:item.enable forKey:@"PushNotification"];
      break;
    case 302: // Force Background
      shareAppDelegate.enableForceBackground = item.enable;
      [settings setBool:item.enable forKey:@"ForceBackground"];
      break;
    }
  }
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  // Return the number of sections.
  return kNumbersOfSections;
}

- (NSInteger)tableView:(UITableView *)tableView
    numberOfRowsInSection:(NSInteger)section {
  switch (section) {
  case 0:
    return [settingsAudioCodec count];
  case 1:
    return [settingsVideoCodec count];
  case 2:
    return [settingsAdvanceFeature count];
  default:
    return 0;
  }
}

- (NSString *)tableView:(UITableView *)tableView
    titleForHeaderInSection:(NSInteger)section {
  switch (section) {
  case 0:
    return kAudioCodecsKey;
  case 1:
    return kVideoCodecsKey;
  case 2:
    return kAdvanceFeatureKey;
  default:
    return nil;
  }
}
- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  SettingCell *cell = (SettingCell *)[tableView
      dequeueReusableCellWithIdentifier:@"settingCell"];

  SettingItem *item = nil;
  switch (indexPath.section) {
  case 0:
    item = [settingsAudioCodec objectAtIndex:indexPath.row];
    break;
  case 1:
    item = [settingsVideoCodec objectAtIndex:indexPath.row];
    break;
  case 2:
    item = [settingsAdvanceFeature objectAtIndex:indexPath.row];
    break;
  default:
    return cell;
  }

  [cell SetItem:item];

  return cell;
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView
    didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  // Navigation logic may go here. Create and push another view controller.
  /*
   <#DetailViewController#> *detailViewController = [[<#DetailViewController#>
   alloc] initWithNibName:@"<#Nib name#>" bundle:nil];
   // ...
   // Pass the selected object to the new view controller.
   [self.navigationController pushViewController:detailViewController
   animated:YES];
   */
}

@end

File: UCSample_swift/UCSample/SettingsViewController.swift
//
//  SettingsViewController.m
//  UCSample
//
//  Copyright (c) 2013 PortSIP Solutions, Inc. All rights reserved.
//

let kNumbersOfSections = 3

let kAudioCodecsKey = "Audio Codecs"
let kVideoCodecsKey = "Video Codecs"
let kAdvanceFeatureKey = "Advance Feature"
class SettingsViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
    var portSIPSDK: PortSIPSDK!

    var settingsAudioCodec: [SettingItem]!
    var settingsVideoCodec: [SettingItem]!
    var settingsAdvanceFeature: [SettingItem]!

    @IBOutlet var tableView: UITableView!

    override func viewDidLoad() {
        super.viewDidLoad()

        settingsAudioCodec = []

        var item = SettingItem()

        item.index = 0
        item.name = "OPUS"
        item.enable = true
        item.codeType = AUDIOCODEC_OPUS.rawValue
        settingsAudioCodec.append(item)

        item = SettingItem()
        item.index = 1
        item.name = "G.729"
        item.enable = true
        item.codeType = AUDIOCODEC_G729.rawValue
        settingsAudioCodec.append(item)

        item = SettingItem()
        item.index = 2
        item.name = "PCMA"
        item.enable = true
        item.codeType = AUDIOCODEC_PCMA.rawValue
        settingsAudioCodec.append(item)

        item = SettingItem()
        item.index = 3
        item.name = "PCMU"
        item.enable = true
        item.codeType = AUDIOCODEC_PCMU.rawValue
        settingsAudioCodec.append(item)

        item = SettingItem()
        item.index = 4
        item.name = "GSM"
        item.enable = false
        item.codeType = AUDIOCODEC_GSM.rawValue
        settingsAudioCodec.append(item)

        item = SettingItem()
        item.index = 5
        item.name = "G.722"
        item.enable = false
        item.codeType = AUDIOCODEC_G722.rawValue
        settingsAudioCodec.append(item)

        item = SettingItem()
        item.index = 6
        item.name = "iLBC"
        item.enable = false
        item.codeType = AUDIOCODEC_ILBC.rawValue
        settingsAudioCodec.append(item)

        item = SettingItem()
        item.index = 7
        item.name = "AMR"
        item.enable = false
        item.codeType = AUDIOCODEC_AMR.rawValue
        settingsAudioCodec.append(item)

        item = SettingItem()
        item.index = 8
        item.name = "AMRWB"
        item.enable = false
        item.codeType = AUDIOCODEC_AMRWB.rawValue
        settingsAudioCodec.append(item)

        item = SettingItem()
        item.index = 9
        item.name = "SpeexNB(8Khz)"
        item.enable = false
        item.codeType = AUDIOCODEC_SPEEX.rawValue
        settingsAudioCodec.append(item)

        item = SettingItem()
        item.index = 10
        item.name = "SpeexWB(16Khz)"
        item.enable = false
        item.codeType = AUDIOCODEC_SPEEXWB.rawValue
        settingsAudioCodec.append(item)

        item = SettingItem()
        item.index = 11
        item.name = "ISACWB(16Khz)"
        item.enable = false
        item.codeType = AUDIOCODEC_ISACWB.rawValue
        settingsAudioCodec.append(item)

        item = SettingItem()
        item.index = 12
        item.name = "ISACSWB(32Khz)"
        item.enable = false
        item.codeType = AUDIOCODEC_ISACSWB.rawValue
        settingsAudioCodec.append(item)

        // Video codec item
        settingsVideoCodec = []
        item = SettingItem()
        item.index = 102
        item.name = "H.264"
        item.enable = true
        item.codeType = VIDEO_CODEC_H264.rawValue
        settingsVideoCodec.append(item)

        item = SettingItem()
        item.index = 103
        item.name = "VP8"
        item.enable = false
        item.codeType = VIDEO_CODEC_VP8.rawValue
        settingsVideoCodec.append(item)

        item = SettingItem()
        item.index = 104
        item.name = "VP9"
        item.enable = false
        item.codeType = VIDEO_CODEC_VP9.rawValue
        settingsVideoCodec.append(item)

        // Advance Feature
        settingsAdvanceFeature = []
        let appDelegate = UIApplication.shared.delegate as! AppDelegate
        item = SettingItem()
        item.index = 300
        item.name = "Integrated Calling"
        item.enable = appDelegate._callManager.enableCallKit
        item.codeType = -1
        settingsAdvanceFeature.append(item)

        item = SettingItem()
        item.index = 301
        item.name = "Push Notification"
        item.enable = appDelegate._enablePushNotification!
        item.codeType = -1
        settingsAdvanceFeature.append(item)

        item = SettingItem()
        item.index = 302
        item.name = "Force Background"
        item.enable = appDelegate._enableForceBackground!
        item.codeType = -1
        settingsAdvanceFeature.append(item)

        tableView.autoresizingMask = UIView.AutoresizingMask.flexibleHeight
        tableView.separatorStyle = UITableViewCell.SeparatorStyle.singleLine
        tableView.delegate = self
        tableView.dataSource = self
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    override func viewWillDisappear(_: Bool) {
        // save change
        portSIPSDK.clearAudioCodec()

        for item in settingsAudioCodec {
            if item.enable {
                portSIPSDK.addAudioCodec(AUDIOCODEC_TYPE(item.codeType))
            }
        }

        portSIPSDK.clearVideoCodec()
        for item in settingsVideoCodec {
            if item.enable {
                portSIPSDK.addVideoCodec(VIDEOCODEC_TYPE(item.codeType))
            }
        }

        let settings: UserDefaults? = UserDefaults.standard
        let appDelegate = UIApplication.shared.delegate as! AppDelegate
        for item in settingsAdvanceFeature {
            switch item.index {
            case 300:
                // Integrated Calling
                if #available(iOS 10.0, *) {
                    appDelegate._callManager.enableCallKit = item.enable
                }

                settings?.set(item.enable, forKey: "CallKit")
            case 301:
                // Push Notification
                if appDelegate._enablePushNotification != item.enable {
                    appDelegate._enablePushNotification = item.enable
                    appDelegate.updatePushStatusToSipServer()
                }
                settings?.set(item.enable, forKey: "PushNotification")
            case 302:
                appDelegate._enableForceBackground = item.enable
                settings?.set(item.enable, forKey: "ForceBackground")
            default:
                break
            }
        }
    }

    func numberOfSections(in _: UITableView) -> Int {
        kNumbersOfSections
    }

    func tableView(_: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch section {
        case 0:
            return settingsAudioCodec.count
        case 1:
            return settingsVideoCodec.count
        case 2:
            return settingsAdvanceFeature.count
        default:
            return 0
        }
    }

    func tableView(_: UITableView, titleForHeaderInSection section: Int) -> String? {
        switch section {
        case 0:
            return kAudioCodecsKey
        case 1:
            return kVideoCodecsKey
        case 2:
            return kAdvanceFeatureKey
        default:
            return ""
        }
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "settingCell") as! SettingCell

        var item: SettingItem
        switch indexPath.section {
        case 0:
            item = settingsAudioCodec[indexPath.row]
        case 1:
            item = settingsVideoCodec[indexPath.row]
        case 2:
            item = settingsAdvanceFeature[indexPath.row]
        default:
            return cell
        }

        cell.SetItem(item)

        return cell
    }

    func tableView(_: UITableView, didSelectRowAt _: IndexPath) {}
}
